### ISSUE TITLE
Database events

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#2586]
I have already asked this question in another issue.

Are there any database events such as ObjectInserting, ObjectInserted, ObjectUpdating, ObjectUpdated, ObjectDeleting, ObjectDeleted?

If no, then what is the best approach to automatically fill CreateDate, ChangeDate values?

**Comment by per-samuelsson** [comment]
@miyconst 

Not sure why this issue never was answered. Sorry.

There are a few callbacks available, some quite stable, others under review and yet others we have supported in earlier versions, but that currently don't work fully.

Let me try summarize:

#### Constructors
Standard **constructors** can be used to initialize instances as they are created. I guess they don't map 1-1 with either ObjectInserting or ObjectInserted, as per your question though.

Example:
```C#
[Database] public class GitHubUser {
  public string Name;
  public string Uri;

  public GitHubUser() : this("johh-doe") {}
  public GitHubUser(string name) {
    Name = name;
    Uri = "https://github.com/" + name;
  }
}
```

Constructors should be safe to use since I very much doubt we'll change how they relate to Starcounter, and they should also be considered stable.

#### IEntity
There is an optional interface called ```IEntity``` with method that will be invoked by Starcounter every time an object of the implementing class is deleted.

Example:
```C#
[Database] public class GitHubUser : IEntity {
  public void OnDelete() {
    Repositories.DeleteAllBelongingTo(this);
  }
}
```

I guess this one is pretty much ObjectDeleting, and as you can see, it can be used to implement manual cascading deletes.

I think ```IEntity``` is fairly stable (at least it was in 1.x, and I don't see any obvious flaws in the code right now). However, it will possibly be refactored and/or extended, but I can't see us removing this kind of functionality completely.

#### Commit hooks
In 1.x, we supported something called *commit hooks*, and they will be reintroduced in some similar fashion pretty soon. I'm quite sure they **dont** work (or maybe doesnt even exist) in the current version of our stable release though.

If you are interested, you can find a lot of information about how they once worked and how they have changed and are implemented in the upcoming version in #2235 and #2339.

**Comment by miyconst** [comment]
@per-samuelsson thank you.

IEntity is right what I need. If only it also contains OnUpdate & OnInsert...
I have already heard about the commit hooks, but they don't work, so I didn't even try to use it.
Will read about the new hooks.
